### ISSUE TITLE
MixPanel consistency

### ISSUE THREAD

**Issue by githubbob42** [githubbob42/mingle2github2#4852]
**[Mingle Card: 5201](https://liquidframeworks.mingle.thoughtworks.com/projects/alpine_mobile/cards/5201)**
Narrative
---------

In mix panel, an individual’s properties currently listed are different in Back office and mobile.

Can we add additional properties so both (Back office and mobile) have the same properties?

FYI - City, County and Region are default properties that are stored in Mix panel libraries. I know LFW uses a server side Mix panel library.

Acceptance Criteria
-------------------

In Back Office, most customers get their name repeated, probably because of spaces in their names or special characters. Fix the code so that all events get reported for the same customer name.

The grid below suggest what needs to be added to each

```
| | |
|-|-|

  
   
   

Properties

   |
   
   

Back office

   |
   
   

Mobile

   |
  
   
   

City

   |
   
   

   

\*

   |
   
   

Needs to be added

   |
  
   
   

Country

   |
   
   

   

\*

   |
   
   

Needs to be added

   |
  
   
   

Region

   |
   
   

\*

   |
   
   

Needs to be added

   |
  
   
   

Email

   |
   
   

\*

   |
   
   

\*

   |
  
   
   

First Name

   |
   
   

\*

   |
   
   

\*

   |
  
   
   

Last Name

   |
   
   

\*

   |
   
   

\*

   |
  
   
   

Last Seen

   |
   
   

\*

   |
   
   

\*

   |
  
   
   

Time Zone

   |
   
   

\*

   |
   
   

   |
  
   
   

Office

   |
   
   

Needs to be added

   |
   
   

\*

   |
  
   
   

Organization

   |
   
   

Needs to be added

   |
   
   

\*

   

   |
  


```

_Analysis_
----------

**Mobile:**

We currently user a nodejs’s mixpanel library to send out analytics events to mixpanel. Because this is a server side library, nodejs by default doesn’t include ip address in the event. On their dashboards, based on the ip provided, mixpanel tacks on City, Country, Region field to each event with configured with an ip address.  We can pull in the client ip address from the req.headers\['x-forwarded-for'\] or req.connection.remoteAddress fields. 

We currently have mixpanel events that are generated on mobile, and sent up to the server, and then mixpanel events generated on the server for specific client actions(login, logout, sync duration, etc..).  Both types of events should have the client’s ip address tacked on.

**Back\_Office triggers:**

We currently have several triggers that publish out events for create events. While we include organization name in each the events, we don’t include the officeName or the ip address(to allow for the auto-populating of region, country, and city fields by mixpanel). The seven triggers are for CreditMemo, Invoice, Payment, Job, PriceBookItem, PriceBook, TIcket when these objects are created. In a corresponding fashion, they generate JobCreated, TicketCreated, CreditMemoCreated, InvoiceCreated, PaymentCreated, PriceBookCreated, PriceBookItemCreated.

Ex:

Because these events are generated inside of a trigger, there’s no way for me to grab the ip address of the user that launched the action that resulted in the trigger getting executed. The only change made will be the adding of office to each of these events.

**Back Office builders:**

We currently publish out mixpanel events for ticketbuilder, quotebuilder, pricebook item builder. These events include location data as well as officeName, and organiziationName. In that sense, they have a complete set of data required by the user story. However, each builder has it’s own naming convention with fields such as org, user info, officeName. Each of the builder supersets were updated to now be consistant with each other. The only negative aspect of this, is that we now have a mountain of old records with mismatched fieldnames on mixpanel.

The back office builders are also passing in user info that has been escaped. This was a bug that was introduced some time ago, but has now resulted in organiziations appearing twice in mixpanel’s historical searches. Below are examples of duplicate records, some with an organization name that is unescaped(Atchafalya 

!clip-6cee-8f29.png!

Related Stories
---------------

Mobile Changes: [https://github.com/Liquidframeworks/alpine-server/pull/209](https://github.com/Liquidframeworks/alpine-server/pull/209)

Back office Changes on Package:  [https://github.com/Liquidframeworks/alpine-package/pull/600](https://github.com/Liquidframeworks/alpine-package/pull/600)

Changs on Back office builders:

1.  **InvoiceBuilder**: [https://github.com/Liquidframeworks/invoicebuilder/pull/20](https://github.com/Liquidframeworks/invoicebuilder/pull/20)
2.  **TicketBuilder: **[https://github.com/Liquidframeworks/ticketbuilder/pull/99](https://github.com/Liquidframeworks/ticketbuilder/pull/99)
3.  **PriceBookBuilder: **[https://github.com/Liquidframeworks/pricebookbuilder/pull/41](https://github.com/Liquidframeworks/pricebookbuilder/pull/41)
4.  **QuoteBuilder: **[https://github.com/Liquidframeworks/quotebuilder/pull/54](https://github.com/Liquidframeworks/quotebuilder/pull/54)

Tasks
-----

{{ table query: SELECT Number, Name, Owner, 'Task Status' WHERE Type = Task and Story = THIS CARD ORDER BY Status }}

Defects
-------

{{ table query: SELECT Number, Name, Owner, 'Status' WHERE Type = Defect and 'Related Story' = THIS CARD }}

Test Plan
---------

The code looks fine, but I don’t see that it is going to add everything we need from Mobile. Please check it over carefully.

**Comment by githubbob42** [comment]
**Pull Request #209**

**Comment by githubbob42** [comment]
Godswill : 07/28/2017 11:07 CDT
Test Result: Passed PR Build: 4.0.1.br710 | 209 User: godswillfield@test4.com/LastPass Hardware: MacBook Pro OS: MacOS 10.12.5 Browser: Version 58.0.3029.110 (64-bit) URL: https://fxqa4.herokuapp.com Test Case: Not yet created. Test note: Mobile: After the server branch was deployed, I was able to retrieve all the location( Country, City and Region) fields on the mixpanel event view when I logged in and sync on mobile. I attached a screenshot of this even on the card. Back office Builders: After applying pull request 209 to server branch, I was able to verify that location data which includes country, city, and region as well as officeName, and organiziationName were displayed on the mixpanel events view for ticketbuilder, quotebuilder, pricebook item builder and invoice item builder. Attached on the card are events screenshots
